HW0/P5/P5.txt

```diff
@@ -1 +1,57 @@
 # Your answers here.
+1.  255 seconds (by whatever method - either by consecutively adding the next
+    bag to a running total, or pairwise, 128 + 64 + 32 + 16 + 8 + 4 + 2 + 1).
+
+2.  Yes - the 8 cashiers can split the bags evenly (so each has 32 bags, which
+    take 31 seconds to add), then four of them each add their total to the
+    total of one of the other four (1 additional second), then two of the
+    remaining four add the remaining two pairs (1 more second), then one
+    cashier adds these results (1 more second), for 34 seconds total.
+
+3.  8 seconds (in the first second, 128 cashiers each add a pair of bags;
+    the next second, 64 add pairs of those results; and so on, until one
+    remaining cashier adds the final total in the 8th second).
+
+4.  Log_2(N) seconds (because the number of sums is halved each second).
+
+5.  15 seconds (in the first second, 128 cashiers each add a pair of bags; the
+    next second, half of those 128 communicate their sum to the other half;
+    those 64 add their sums to the sums they were just told; and so on,
+    requiring one additional second of communication after each one-second
+    round of the instantaneous-communication version, except the final round).
+
+6.  If each cashier takes 1 bag, 260 seconds (the last bag is handed to a
+    cashier in the 256th second; prior to that, each bag handed out is
+    communicated to one of two cashiers, so one is always adding while the
+    other gets a number from the cashier that just received a bag. When the
+    last bag is handed out, one cashier has just added the 127th bag to their
+    total, while the other cashier has just been told their 128th number; in
+    the 257th second, the last bag's number is communicated to the former
+    cashier, and the latter cashier adds their 128th value to their running
+    total; in the 258th second, the former cashier adds the last number to
+    their total; in the 259th second, one communicates their total to the
+    other; and in the 260th second, they add their totals).
+
+    If each cashier takes 2 bags, 133 seconds (the last bags are handed to a
+    cashier in the 128th second; one second later, the last cashier to receive
+    a bag adds their two bags together; and from that point, as above, an
+    additional four seconds are necessary to reach the final total).
+
```

Comment:
Yes, in the everyone-takes-n-bags-from-me-directly scenario, I'm pretty sure this is the best you can do - each cashier adds up their own n bags and then tells the total to a cashier who isn't currently adding (and since only one of the first two cashiers to receive their bags is adding at a time, because they're offset by one second, this means that there's no advantage - at least, not that I could find - to having more than two cashiers collecting the totals). It's still much less efficient than some methods that don't require all the cashiers to come up to me and take bags directly, though.